examples/traffic_histories_to_observations.py

```diff
@@ -1,130 +1,330 @@
-                except ControllerOutOfLaneException:
-                    logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
-                    vehicles_off_road.add(veh_id)
+        self._smarts.destroy()
 
-            valid_vehicles = {v for v in history_vehicles if v not in vehicles_off_road}
-            obs, _, _, dones = smarts.observe_from(list(valid_vehicles))
-            _record_data(smarts.elapsed_sim_time, obs, collected_data)
+    def _record_data(
+        self,
+        collected_data,
+        current_vehicles,
+        off_road_vehicles,
+        selected_vehicles,
+    ):
+        # Attach sensors to each vehicle
+        valid_vehicles = (current_vehicles - off_road_vehicles) & selected_vehicles
+        for veh_id in valid_vehicles:
+            try:
+                self._smarts.attach_sensors_to_vehicles(self.agent_interface, {veh_id})
+            except ControllerOutOfLaneException:
+                self._logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
+                off_road_vehicles.add(veh_id)
 
-        # an example of how we might save the data per car
-        observation_folder = "collected_observations"
-        if not os.path.exists(observation_folder):
-            os.makedirs(observation_folder)
-        for car, data in collected_data.items():
-            outfile = f"{observation_folder}/{scenario.name}_{scenario.traffic_history.name}_{car}.pkl"
-            with open(outfile, "wb") as of:
-                pickle.dump(data, of)
+        # Get observations from each vehicle and record them
+        obs = dict()
+        obs, _, _, _ = self._smarts.observe_from(list(valid_vehicles))
+        resolutions = {}
+        for id_ in list(obs):
+            if obs[id_].top_down_rgb:
+                resolutions[id_] = obs[id_].top_down_rgb.metadata.resolution
+            ego_state = obs[id_].ego_vehicle_state
+            if ego_state.lane_index is None:
+                del obs[id_]
+            else:
+                mission = self._missions[ego_state.id]
+                if mission:
+                    # doh! ego_state is immutable!
+                    new_ego_state = ego_state._replace(mission=mission)
+                    obs[id_] = replace(obs[id_], ego_vehicle_state=new_ego_state)
+        # TODO: handle case where neighboring vehicle has lane_index of None too
+        t = self._smarts.elapsed_sim_time
+        for car, car_obs in obs.items():
+            collected_data.setdefault(car, {}).setdefault(t, {})
+            collected_data[car][t] = car_obs
 
-    smarts.destroy()
+        if not self._output_dir:
+            return
+
+        # Write top-down RGB image to a file for each vehicle if we have one
+        for agent_id, agent_obs in obs.items():
+            if agent_obs.top_down_rgb is not None:
+                rgb_data = agent_obs.top_down_rgb.data
+                h, w = rgb_data.shape[0], rgb_data.shape[1]
```

Comment:
Numpy arrays unpack.

```suggestion
                h, w = rgb_data.shape
```